File: 141-hasCycle.go
// Package main provides ...
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

/**
 * 链表有环
 */
func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}

	slow := head
	fast := head.Next

	for fast != slow {
		if fast == nil || fast.Next == nil {
			return false
		}

		slow = slow.Next
		fast = fast.Next.Next
	}
	return true
}

func detectCycle(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return nil
	}

	slow := head
	fast := head.Next

	for fast != nil && fast.Next != nil {
		if fast == slow {
			slow = head
			break
		}

		slow = slow.Next
		fast = fast.Next.Next
	}

	if fast == nil || fast.Next == nil {
		return nil
	}

	for fast != slow {
		slow = slow.Next
		fast = fast.Next
	}

	return fast
}

func NewNode(val int) *ListNode {
	return &ListNode{val, nil}
}

func main() {
	head := NewNode(3)
	cur := head
	cur.Next = NewNode(2)
	cur = cur.Next
	cur.Next = NewNode(0)

	cur = cur.Next
	cur.Next = NewNode(-4)

	r := detectCycle(head)
	fmt.Println(r)
}

File: 141-hasCycle.php
<?php
require_once __DIR__ . '/structure/linklist.php';

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */

class Solution {
    /**
     * @param ListNode $head
     * @return ListNode
     */
    function detectCycle($head)
    {
       if (!$head || !$head->next) {
           return null; 
       } 

       $slow = $fast = $head;
       $flag = 0;
       while ($fast->next->next) {
           $slow = $slow->next;
           $fast = $fast->next->next;

           if ($fast == $slow) {
               $flag = 1;
               break; 
           }
       }
       if (!$flag) {
          return null; 
       }

       for ($slow = $head; $fast != $slow; $slow=$slow->next,$fast =$fast->next)
           return $fast;

       return null;
    }
}

//$head = new ListNode(1);
//insert($head, 2);
//insert($head, 0);
//insert($head, -4);
//insert($head, -3);
//insert($head, -5);
//insert($head, -6);
//$circle = $head->next->next->next;
//$head->next->next->next->next->next->next = $circle;
//lPrint($head);
//die;
//
$head = new ListNode(3);
insert($head, 2);
insert($head, 0);
insert($head, -4);

$r = (new Solution())->detectCycle($head);
print_r($r);
